packages/client-http/src/fetch-util.ts

```diff
@@ -0,0 +1,322 @@
+   * @param options.initialTokens - the initial number of tokens, can be larger than `maxTokens`
+   * @returns the builder itself
+   */
+  rateLimit(tokenFillRate: number, { maxTokens = tokenFillRate, initialTokens = tokenFillRate } = {}) {
+    let lastRefillTime = Number.NEGATIVE_INFINITY;
+    let tokens = initialTokens;
+    let nextRun = Promise.resolve();
+
+    return this.compose(next => request => {
+      nextRun = nextRun
+        // catch the potential abort of the previous run
+        .catch(() => {})
+        .then(async () => {
+          refillTokens();
+          // we can only send a request if we have at least one token
+          if (tokens < 1) {
+            // wait until tokens is one
+            await abortableSleep(((1 - tokens) / tokenFillRate) * 1000, request.signal);
+          }
+        });
+      return nextRun.then(() => {
+        tokens--;
+        return next(request);
+      });
+    });
+
+    function refillTokens() {
+      const currentTime = Date.now();
+      // initialTokens might be higher than maxTokens, so we don't refill until we've dropped below maxTokens
+      if (Number.isFinite(lastRefillTime) && tokens < maxTokens) {
+        const elapsed = (currentTime - lastRefillTime) / 1000;
+        tokens = Math.min(maxTokens, tokens + elapsed * tokenFillRate);
+      }
+      lastRefillTime = currentTime;
+    }
+  }
+
+  /**
+   * Retry rejected requests.
+   *
+   * @param options.maxRetries - the maximum number of retires, defaults to unbounded
+   * @param options.delay - the delay between retries, defaults to zero
+   * @param options.backoff - multiplier for exponentially increasing the delay, defaults to `2`
+   * @param options.jitter - random jitter as a ratio (.i.e `0.1` means ±10% jitter), defaults to zero
+   * @param options.maxDelay - the maximum duration at which backoff will no longer apply, defaults to unbounded
+   * @returns the builder itself
+   */
+  retry({
+    maxRetries = Number.POSITIVE_INFINITY,
+    delay = 0,
+    maxDelay = Number.POSITIVE_INFINITY,
+    backoff = 2,
+    jitter = 0,
+  } = {}): this {
+    return this.compose(next => async request => {
+      let retryCount = 0;
+
+      const doRetry = async (e: unknown): Promise<Response> => {
+        // if there are no more attempts we throw the last error
+        if (retryCount >= maxRetries) throw e;
+
+        const jitterFactor = 1 + 2 * Math.random() * jitter - jitter;
+        await abortableSleep(jitterFactor * Math.min(maxDelay, delay * Math.pow(backoff, retryCount)), request.signal);
+        retryCount++;
+        return next(request.clone()).catch(doRetry);
+      };
+
+      return next(request.clone()).catch(doRetry);
+    });
+  }
+
+  /**
+   * Modify the request.
+   *
+   * @param mod - a (possibly async) function to transform the request
+   * @returns the builder itself
+   */
+  modifyRequest(mod: (request: Request) => Promise<Request> | Request): this {
```

Comment:
Why do we need a modifyRequest when we already have compose? Is it because of the async funcationality?

Reply:
It's just a simpler function intended for use of the API. `compose` is private.